practice/prototype_anc_samp.py

```diff
@@ -0,0 +1,624 @@
+        new_mutations = [(pos, new_mutation_node_id, generation)
+                         for pos in new_mutation_locs]
+        mutations_all.extend(new_mutations)
+    return mutations_all
+
+
+def wf(N, simplifier, tracker, recrate, murate, anc_sample_gen, ngens):
+    """
+    For N diploids, the diploids list contains 2N values.
+    For the i-th diploids, diploids[2*i] and diploids[2*i+1]
+    represent the two chromosomes.
+
+    We simulate constant N here, according to a Wright-Fisher
+    scheme:
+
+    1. Pick two parental indexes, each [0,N-1]
+    2. The two chromosome ids are diploids[2*p] and diploids[2*p+1]
+       for each parental index 'p'.
+    3. We pass one parental index on to each offspring, according to
+       Mendel, which means we swap parental chromosome ids 50% of the time.
+    4. Crossing over is a Poisson process, and the code used here (above)
+       is modeled after that in our C++ implementation.
+    5. Mutation is a Poisson process, mutations are added according to each
+       parental index. 
+    """
+    diploids = np.arange(2 * N, dtype=np.uint32)
+    # so we pre-allocate the space. We only need
+    # to do this once b/c N is constant.
+    tracker.nodes = np.empty([0], dtype=node_dt)
+    tracker.mutations = np.empty([0], dtype=mutation_dt)
+
+    mutation_lookup = dict()
+
+    ancestral_gen_counter = 0
+    next_id = len(diploids)  # This will be the next unique ID to use
+    assert(max(diploids) < next_id)
+    for gen in range(ngens):
+        ancestral_samples = []
+        # Let's see if we will do some GC:
+        gc_rv = simplifier(gen, tracker)
+        # If so, let the tracker clean up:
+        tracker.post_gc_cleanup(gc_rv)
+        if gc_rv[0] is True:
+            assert(len(tracker.nodes) == 0)
+            assert(len(tracker.edges) == 0)
+            assert(len(tracker.mutations) == 0)
+
+            # Reset our mutation lookup via a dictionary comprehension
+            mutation_lookup = {i: True for i in simplifier.sites.position}
+
+            # If we did GC, we need to reset
+            # some variables.  Internally,
+            # when msprime simplifies tables,
+            # the 2N tips are entries 0 to 2*N-1
+            # in the NodeTable, hence the
+            # re-assignment if diploids.
+            # We can also reset next_id to the
+            # length of the current NodeTable,
+            # keeping risk of integer overflow
+            # to a minimum.
+            next_id = int(gc_rv[1])
+            diploids = np.arange(2 * N, dtype=np.uint32)
+
+        # Empty offspring list.
+        # We also use this to mark the "samples" for simplify
+        new_diploids = np.empty([len(diploids)], dtype=diploids.dtype)
+
+        if(ancestral_gen_counter < len(anc_sample_gen) and gen + 1 == anc_sample_gen[ancestral_gen_counter][0]):
+            assert(anc_sample_gen[ancestral_gen_counter][1] < (2 * N))
+            ancestral_samples = np.random.choice(
```

Comment:
Should this be `ancestral_samples.extend(...)`.  It seems possible to over-write your ancestral samples if you get them more often than you simplify?

Reply:
Good point! I haven't tested that in a sample yet, but yes I'm pretty sure that would happen if I did.

Reply:
Actually I just remembered, this isn't an issue because I update the tracker every generation, adding the new ancestral samples to be taken in the next (and all subsequent) gc to the ARG tracker.

Reply:
One thing that does seem to happen with multiple ancestral samples per gc, is that the regular sample nodes and the ancestral sample nodes seem to get mixed up together which doesn't happen with 1 or fewer ancestral samples per gc. This doesn't affect the actual simulation, but my original method of doing the final sampling and simplification relied on the ancestral samples being first in the node table. I have now amended it and added a fix so that it will still take ancestral samples even if the population size is very small.

Reply:
Is it possible that this is related to the number of mutations issue?

Reply:
Sadly, no (or at least if so, it represents a problem above and beyond what I've fixed so far). The above issue only happened on simulations with small population sizes and only affected the final simplification. I have pushed a fix and the number of mutations issue is still happening. 

It's possible that there is something related to this that I'm missing, but I can't see it yet.